File: extensions/postgres-flyway/src/main/java/de/sovity/edc/extension/postgresql/FlywayMigrator.java
package de.sovity.edc.extension.postgresql;

import com.zaxxer.hikari.HikariDataSource;
import de.sovity.edc.utils.config.CeConfigProps;
import lombok.val;
import org.eclipse.edc.spi.monitor.Monitor;
import org.eclipse.edc.spi.persistence.EdcPersistenceException;
import org.flywaydb.core.api.FlywayException;

public class FlywayMigrator {

    private final PostgresFlywayConfig config;
    private final HikariDataSource dataSource;
    private final Monitor monitor;
    private final FlywayFactory flywayFactory;

    public FlywayMigrator(PostgresFlywayConfig config, HikariDataSource dataSource, Monitor monitor) {
        this.config = config;
        this.dataSource = dataSource;
        this.monitor = monitor;
        this.flywayFactory = new FlywayFactory(config);
    }

    public void updateDatabaseWithLegacyHandling() {
        cleanDatabase();
        updateDatabase();
    }

    private void cleanDatabase() {

        boolean shouldClean = config.flywayClean();
        boolean canClean = config.flywayCleanEnabled();
        if (shouldClean) {
            if (!canClean) {
                throw new IllegalStateException("In order to clean the history both %s and %s must be set to true.".formatted(
                        CeConfigProps.EDC_FLYWAY_CLEAN.getProperty(), CeConfigProps.EDC_FLYWAY_CLEAN_ENABLE.getProperty()
                ));
            }
            monitor.info(() -> "Cleaning database before migrations, since %s=true and %s=true.".formatted(
                CeConfigProps.EDC_FLYWAY_CLEAN.getProperty(), CeConfigProps.EDC_FLYWAY_CLEAN_ENABLE.getProperty()
            ));

            val flyway = flywayFactory.setupFlywayForUnifiedHistory(dataSource);
            flyway.clean();
        }
    }

    private void updateDatabase() {
        val flyway = flywayFactory.setupFlywayForUnifiedHistory(dataSource);

        try {
            flyway.migrate();
        } catch (FlywayException e) {
            throw migrationFailed(e);
        }
    }

    private EdcPersistenceException migrationFailed(Exception e) {
        return new EdcPersistenceException("Flyway migration failed", e);
    }
}

File: connector/ce/connector-ce/src/main/kotlin/de/sovity/edc/ce/modules/db/flyway/FlywayMigrator.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.ce.modules.db.flyway

import de.sovity.edc.ce.config.CeConfigProps.SOVITY_FLYWAY_CLEAN
import de.sovity.edc.ce.config.CeConfigProps.SOVITY_FLYWAY_CLEAN_ENABLE
import org.eclipse.edc.spi.monitor.Monitor
import org.eclipse.edc.spi.persistence.EdcPersistenceException
import org.eclipse.edc.spi.system.configuration.Config
import org.flywaydb.core.Flyway
import org.flywaydb.core.api.FlywayException

class FlywayMigrator(
    private val config: Config,
    private val monitor: Monitor
) {
    fun cleanAndMigrate(flyway: Flyway) {
        cleanDatabase(flyway)
        migrate(flyway)
    }

    private fun cleanDatabase(flyway: Flyway) {
        val shouldClean = SOVITY_FLYWAY_CLEAN.getBooleanOrFalse(config)
        val canClean = SOVITY_FLYWAY_CLEAN_ENABLE.getBooleanOrFalse(config)
        if (shouldClean) {
            check(canClean) {
                "In order to clean the history both $SOVITY_FLYWAY_CLEAN and $SOVITY_FLYWAY_CLEAN_ENABLE must be set to true."
            }
            monitor.info("Cleaning database before migrations, since $SOVITY_FLYWAY_CLEAN=true and $SOVITY_FLYWAY_CLEAN_ENABLE=true")
            flyway.clean()
        }
    }

    private fun migrate(flyway: Flyway) {
        try {
            flyway.migrate()
        } catch (e: FlywayException) {
            throw EdcPersistenceException("Flyway migration failed", e)
        }
    }
}
